JAVA-657: Debounce control connection queries.

adutra opened the pull request:
Also contains a fix for JAVA-784.

Commit:
JAVA-657: Debounce control connection queries.

Includes squashed contributions from olim7t and tolbertam.

adutra commented on driver-core/src/main/java/com/datastax/driver/core/AbstractEventDebouncer.java:
```diff
@@ -0,0 +1,212 @@
+        logger.trace("Debouncer started");
+        eventsLock.lock();
+        try {
+            if (!events.isEmpty()) {
```
This is basically an anticipation of JAVA-784.

adutra commented on driver-core/src/main/java/com/datastax/driver/core/AbstractEventDebouncer.java:
```diff
@@ -0,0 +1,212 @@
+                try {
+                    // if the queue is full and the debouncer is not started,
+                    // we have no other choice than to block and wait
+                    awaitUntilStarted();
```
again, relates to JAVA-784.

adutra commented on driver-core/src/main/java/com/datastax/driver/core/AbstractEventDebouncer.java:
```diff
@@ -0,0 +1,212 @@
+    private void enqueueEvent(T event) {
+        eventsLock.lock();
+        try {
+            while (!events.offer(event)) {
```
Events will accumulate until the queue is full, in which case the delivery is scheduled on the calling thread (back pressure).

adutra commented on driver-core/src/main/java/com/datastax/driver/core/Cluster.java:
```diff
@@ -2043,8 +2084,11 @@ public void run() {
+                        if (refreshSchema) {
+                            submitSchemaRefresh(keyspace, table);
+                            // force immediate event delivery
+                            schemaRefreshRequestDebouncer.deliverEvents();
```
This tries to mimic the old behavior where the refresh was done synchronously in the calling thread.

adutra commented on driver-core/src/main/java/com/datastax/driver/core/Cluster.java:
```diff
@@ -2282,6 +2285,195 @@ public void run() {
+
+            @Override
+            public void deliver(final List<SchemaRefreshRequest> events) {
+                executor.submit(new ExceptionCatchingRunnable() {
```
All the current 3 callbacks do the actual job on the main executor, never on the current thread (=scheduled executor thread or I/O thread).

adutra commented on driver-core/src/main/java/com/datastax/driver/core/Cluster.java:
```diff
@@ -2282,6 +2285,195 @@ public void run() {
+                                @Override
+                                public void runMayThrow() throws InterruptedException, ExecutionException {
+                                    if (controlConnection.refreshNodeInfo(host)) {
+                                        triggerOnUp(host);
```
Previously we would call onUp() and execute the task synchronously; I believe this is not desired because this means the (single) thread of the scheduled executor. So I changed that for the (asynchronous) triggerOnUp().

adutra commented on driver-core/src/main/java/com/datastax/driver/core/Cluster.java:
```diff
@@ -2282,6 +2285,195 @@ public void run() {
+                    // so that we have up-to-date infos on that host before adding it (so we typically
+                    // catch that an upgraded node uses a new cassandra version).
+                    switch (eventType) {
+                        case UP:
```
UP and ADDED events are delayed by 1 second. Previously only ADDED events would be delayed.

adutra commented on driver-core/src/main/java/com/datastax/driver/core/Cluster.java:
```diff
@@ -2282,6 +2285,195 @@ public void run() {
+                                @Override
+                                public void runMayThrow() throws InterruptedException, ExecutionException {
+                                    if (controlConnection.refreshNodeInfo(host)) {
+                                        triggerOnAdd(host);
```
Previously we would call onAdd() and execute the task synchronously; I believe this is not desired because this means the (single) thread of the scheduled executor. So I changed that for the (asynchronous) method triggerOnAdd().

adutra commented on driver-core/src/main/java/com/datastax/driver/core/ControlConnection.java:
```diff
@@ -255,11 +255,14 @@ private Connection tryConnect(Host host, boolean isInitialConnection) throws Con
+            // We need to refresh the node list again;
+            // We want that because the token map was not properly initialized by the first call above,
+            // since it requires the list of keyspaces to be loaded.
+            refreshNodeListAndTokenMap(connection, cluster, false, true);
```
This was called from inside refreshSchema in certain situations; that had to be pulled out of that method; when initiating the connection this method was effectively always called, so we simple call it here.

adutra commented on driver-core/src/main/java/com/datastax/driver/core/ControlConnection.java:
```diff
@@ -286,7 +289,11 @@ public void refreshSchema(String keyspace, String table) throws InterruptedExcep
+            refreshSchema(c, keyspace, table, cluster);
+            // If the table is null, we either rebuild all from scratch or have an updated keyspace. In both cases, rebuild the token map
+            // since some replication on some keyspace may have changed
+            if (table == null)
```
Same remark as above, but here the node list refresh can be done asynchronously.

adutra commented on driver-core/src/test/java/com/datastax/driver/core/CaseSensitivityTest.java:
```diff
@@ -36,12 +37,21 @@ private void assertNotExists(CCMBridge.CCMCluster c, String name) {
             s.execute(String.format(CREATE_KEYSPACE_SIMPLE_FORMAT, ksName, 1));
 
+            // give some time for schema events to be debounced
+            Thread.sleep(500);
```
Has to add lots of these to make the driver pause until the events are debounced :( - even setting intervals to zero is not enough.

adutra commented on driver-core/src/main/java/com/datastax/driver/core/AbstractEventDebouncer.java:
```diff
@@ -0,0 +1,212 @@
+        scheduleNewDelivery();
+    }
+
+    void deliverEvents() throws InterruptedException {
```
The normal case is to have this method called by a thread in the scheduled executor. In case of back pressure, the following threads are also likely to execute this method:
* IO thread (nio worker) in Cluster.handle(Response), if the queue is full (is it a problem?)
* Client thread or cluster-reconnection pool in ControlConnection.tryConnect(Host, boolean) (only happens at initial connection attempt)
* cluster-worker (main executor) in maybeRefreshSchemaAndSignal()

adutra commented on driver-core/src/main/java/com/datastax/driver/core/AbstractEventDebouncer.java:
```diff
@@ -0,0 +1,198 @@
+        }
+    }
+
+    private void deliverEvents() throws InterruptedException {
```
The normal case is to have this method called by a thread in the scheduled executor. In case of back pressure, the following threads are also likely to execute this method:
* IO thread (nio worker) in Cluster.handle(Response), if the queue is full (is it a problem? Note that the calling thread only executes the queue drain, once the events are drained they are processed in the main executor)
* Client thread or cluster-reconnection pool in ControlConnection.tryConnect(Host, boolean) (only happens at initial connection attempt)

adutra commented on driver-core/src/main/java/com/datastax/driver/core/AbstractEventDebouncer.java:
```diff
@@ -0,0 +1,198 @@
+                try {
+                    // if the queue is full and the debouncer is not ready,
+                    // we have no other choice than to block and wait
+                    started.await();
```
This can block an IO thread.

Commit:
JAVA-784: LoadBalancingPolicy.distance() called before init().

adutra commented on driver-core/src/main/java/com/datastax/driver/core/Cluster.java:
```diff
@@ -2043,8 +2117,16 @@ public void run() {
+                        if (refreshSchema) {
+                            submitSchemaRefresh(keyspace, table);
+                            // force synchronous schema refresh
+                            Uninterruptibles.getUninterruptibly(schemaRefreshRequestDebouncer.deliverEvents());
```
This mimics the old, synchronous behavior where ControlConnection.refreshSchema was called directly.

olim7t replied:
Actually the previous version performed the refresh with the _current_ connection (the one that just executed the schema-altering statement), while the debouncer will always do it with the control connection.
The problem is that the control host might not yet be up to date with the schema changes, so the metadata is still stale when the statement completes.
I'm not quite sure how to solve that yet.

adutra commented on driver-core/src/main/java/com/datastax/driver/core/ControlConnection.java:
```diff
@@ -258,11 +257,14 @@ private Connection tryConnect(Host host, boolean isInitialConnection) throws Con
+            // We need to refresh the node list again;
+            // We want that because the token map was not properly initialized by the first call above,
+            // since it requires the list of keyspaces to be loaded.
+            refreshNodeListAndTokenMap(connection, cluster, false, true);
```
This was called from inside refreshSchema in certain situations; that had to be pulled out of that method; when initiating the connection this method was effectively always called, so we simply call it here (synchronously because we need the information ready when the method returns).

olim7t replied:
OK

adutra commented on driver-core/src/main/java/com/datastax/driver/core/ControlConnection.java:
```diff
@@ -289,7 +291,13 @@ public void refreshSchema(String keyspace, String table) throws InterruptedExcep
+            // If the table is null, we either rebuild all from scratch or have an updated keyspace. In both cases, rebuild the token map
+            // since some replication on some keyspace may have changed
+            if (table == null) {
+                cluster.submitNodeListRefresh();
```
Same here, but I'd rather submit a request and flush pending events because this time it's not the first connection attempt; this is called from maybeRefreshSchemaAndSignal().

olim7t replied:
OK

adutra commented on driver-core/src/main/java/com/datastax/driver/core/EventDebouncer.java:
```diff
@@ -0,0 +1,208 @@
+        return started.getCount() == 0 && !stopped;
+    }
+
+    ListenableFuture<?> deliverEvents() {
```
The normal case is to have this method called by a thread in the scheduled executor, when the timer times out. 
But the following threads are also likely to execute this method in rare situations:
- IO thread (nio worker) in Cluster.handle(Response), if the queue is full (back pressure - note that the calling thread only executes the queue drain, once the events are drained they are processed asynchronously in the main executor)
- Client thread or cluster-reconnection pool in ControlConnection.tryConnect(Host, boolean) (only happens at connection attempts, and this method already has blocking operations and everything is done synchronously)
- Cluster main executor in maybeRefreshSchemaAndSignal() - this is the same executor that executes debounced events.

adutra commented on driver-core/src/main/java/com/datastax/driver/core/EventDebouncer.java:
```diff
@@ -0,0 +1,219 @@
+ *
+ * The events are only delivered:
+ * - if the current timer times out, in which case the events are delivered asynchronously;
+ * - if the queue is full, in which case the events are delivered synchronously.
```
The docs do not reflect the current state of the code anymore.

olim7t commented on driver-core/src/main/java/com/datastax/driver/core/EventDebouncer.java:
```diff
@@ -0,0 +1,242 @@
+         *
+         * @param events the events to deliver
+         */
+        ListenableFuture<?> deliver(List<T> events);
```
btw, this could be void because the calling code does nothing with this future.

tolbertam commented on driver-core/src/main/java/com/datastax/driver/core/Cluster.java:
```diff
@@ -1259,6 +1263,28 @@ synchronized void init() {
+
+                        // create debouncers - at this stage, they are not running yet
+                        QueryOptions queryOptions = configuration.getQueryOptions();
+                        this.nodeListRefreshRequestDebouncer = new EventDebouncer<NodeListRefreshRequest>(
```
I think it will be useful, at least for my local stress testing, to track the size of pending events (basically a Gauge over EventDebouncer#eventCount) so I'll add and register a gauge for metrics.  Is it useful enough to include in our code?   Only thing I can think of is to measure impact when tuning with diifferent intervals.

adutra replied:
I was precisely worried about this request showing up one day :) because this counter is not meant to be 100% accurate (see my previous comments). I'm not against the idea of the gauge, but we should put a warning about accuracy on it (it's not atomically updated in respect to the queue so it could temporarily be above the maximum or below zero).

tolbertam replied:
+1.  I think it not entirely being accurate is ok as I think it will only mostly be used for tuning (what intervals to use and such) and for checking to make sure nothing is 'stuck'.

tolbertam commented on driver-core/src/main/java/com/datastax/driver/core/Cluster.java:
```diff
@@ -2082,35 +2129,16 @@ public void handle(Message.Response response) {
-                                }, NEW_NODE_DELAY_SECONDS, TimeUnit.SECONDS);
-                            }
+                            if (newHost != null)
+                                submitNodeRefresh(newHost, HostEvent.ADDED);
```
Nice, eliminates a lot of code duplication :)

tolbertam commented on driver-core/src/main/java/com/datastax/driver/core/Cluster.java:
```diff
@@ -2282,6 +2288,230 @@ public void run() {
+                            }
+                        });
+                    }
+                }, NEW_NODE_DELAY_SECONDS, TimeUnit.SECONDS);
```
This could be somewhat problematic for our tests since we set a really large NEW_NODE_DELAY_SECONDS, though maybe we should stop doing that.

adutra replied:
Wasn't it already so before? The only difference is that the scheduling was done inside the `handle()` method.

tolbertam replied:
Ah, right you are!

tolbertam commented on driver-core/src/main/java/com/datastax/driver/core/Cluster.java:
```diff
@@ -2282,6 +2288,230 @@ public void run() {
+            SchemaRefreshRequest coalesce(SchemaRefreshRequest that) {
+                if(this.keyspace == null || that.keyspace == null)
+                    return new SchemaRefreshRequest(null, null);
+                if(!this.keyspace.equals(that.keyspace))
```
One small concern I have here is that if you have changes to two keyspaces in this interval refreshing the entire schema could be pretty costly, as chances are if you have two keyspaces altered within a small interval that you have a pretty complex schema.   The same logic can be applied to tables within the same keyspace, although I think the way this currently works is reasonable.  We could have some kind of threshold where we decide if X keyspace/tables have changed only then refresh the entire schema/keyspace, otherwise refresh each individually, though don't know how important it is to have something like that.

tolbertam commented on driver-core/src/main/java/com/datastax/driver/core/QueryOptions.java:
```diff
@@ -42,10 +42,32 @@
+
+    public static final int DEFAULT_REFRESH_NODE_INTERVAL_MILLIS = 1000;
+
+    public static final int DEFAULT_REFRESH_SCHEMA_INTERVAL_MILLIS = 1000;
```
Is this default for refresh schema interval too aggressive?  For example, if doing a schema migration, you may be changing a number of tables/keyspaces at a time.  If doing this responsibility, you'd wait for schema agreement after each change, so you could be looking at ~10 seconds between operation, in which case the debouncing will not be very helpful.

olim7t replied:
I think debouncing is mainly intended for when you don't do things responsibly :)

But I agree that the default are open for discussion, we should discuss them once we've run a few more tests.

tolbertam commented on driver-core/src/test/java/com/datastax/driver/core/EventDebouncerIntegrationTest.java:
```diff
@@ -0,0 +1,135 @@
+     * @jira_ticket JAVA-784
+     * @since 2.0.11
+     */
+    @Test(groups = "short")
```
This should probably be marked 'long' since it doesn't use PerClassSingleNodeCluster/takes node down/etc, although it runs decently quickly (~25 secs in my environment).

olim7t replied:
Done.

tolbertam commented on driver-core/src/main/java/com/datastax/driver/core/ControlConnection.java:
```diff
@@ -289,7 +291,13 @@ public void refreshSchema(String keyspace, String table) throws InterruptedExcep
+            // since some replication on some keyspace may have changed
+            if (table == null) {
+                cluster.submitNodeListRefresh();
+                cluster.nodeListRefreshRequestDebouncer.deliverEvents().get();
```
This can get weird since we could be inside a SchemaRefreshRequestDeliveryCallback which is already executing in Cluster.executor, this will submit another task to Cluster.executor.  This gets even weirder if the schemaRefresh was triggered by schemaRefreshRequestDebouncer.deliverEvents in Cluster.maybeRefreshSchemaAndSignal.  This leaves us with a situation where we have:

```
a. A response submitting SchemaRefreshDeiveryCallback in Cluster.executor via maybeRefreshSchemaAndSingal
    b. SchemaRefreshDeliveryCallback submitting Runnable to Cluster.executor:
        c. Runnable calling controlConnection.refreshSchema which submits a nodeListRefresh
            d.  NodeListRefreshDeliveryCallback submitting a Runnable to refreshNodeListAndTokenMap in Cluster.Executor.
```

If Cluster.executor only has 2 threads (dual core or less system or configured that way), wouldn't that starve the executor since in a and c we are waiting on completion of whatever is being submitted at that time?

tolbertam replied:
Was able to reproduce this case by creating a keyspace when NON_BLOCKING_EXECUTOR_SIZE is 2.  The future never completes (it works if NON_BLOCKING_EXECUTOR_SIZE is 3).  Added test case and analysis to this [gist](https://gist.github.com/tolbertam/8c024d5e6fbae1401bea).

adutra replied:
You are absolutely right, we probably won't be able to block here waiting on this future. Conceptually, I think it is acceptable; but the metadata state will not always be accurate and we could have some tests failing more or less deterministically.

olim7t replied:
I don't think we need to wait on the debouncer here: if the token map is stale for a while that's not a big deal, so we don't need to wait synchronously.
I would simply submit the refresh and carry on. I'm working on some changes, I'll include that.

tolbertam replied:
Your changes fix it, thanks!  Even works for NON_BLOCKING_EXECUTOR_SIZE 1 :)

tolbertam committed:
Add tests for query debouncing (JAVA-657).

tolbertam commented on driver-core/src/main/java/com/datastax/driver/core/Cluster.java:
```diff
@@ -2282,6 +2289,234 @@ public void run() {
+                    HostEvent eventType = entry.getValue();
+                    switch (eventType) {
+                        case UP:
+                            schedule(hostUp(host));
```
Shouldn't hostUp be executed right away? It looks like previously only 'schedule' was used for added.

tolbertam replied:
Don't we also have a problem with this sequence of events:

```
onAdd
onUp
```

The UP event will be processed, but we never get an ADDED event.   I think in the case of both 'ADDED' and 'REMOVED' we should not suppress these events if there are newer ones.

tolbertam replied:
It looks like that is the case, however it seems that the only impact of this is that since `onAdd` is never called all the things that happen in there (lbp.onAdd, StateListener.onAdd, ControlConnection.onAdd).  I suppose the same would be the case for onRemove.

Here's some test code to produce this scenario, adding some log statements in NodeRefreshFrequestDeliveryCallback helps show what happens:

``` java
    @Test(groups="short")
    public void should_call_onAdd_with_bootstrap_stop_start() {
        QueryOptions queryOptions = new QueryOptions().setRefreshNodeIntervalMillis(10000);
        CCMBridge ccm = CCMBridge.builder("test").withNodes(1).build();
        Cluster cluster = spy(Cluster.builder()
            .addContactPoint(CCMBridge.ipOfNode(1))
            .withQueryOptions(queryOptions)
            .build());

        try {
            cluster.connect();
            Host.StateListener listener = mock(Host.StateListener.class);
            cluster.register(listener);
            ccm.bootstrapNode(2);
            ccm.stop(2);
            ccm.start(2);
            // TODO: Ensure onAdd and onUp are called on listener.  Sleep for a while now to observe behavior in logs.
            Uninterruptibles.sleepUninterruptibly(100, TimeUnit.SECONDS);
        } finally {
            cluster.close();
            ccm.remove();
        }
    }
```

There are probably a number of ways to approach this, but the way I think this should work is that on NodeRefreshRequestDeliveryCallback:
1. retain at least 1 of each of [REMOVED, ADDED] if they exist.
2. retain the most recent (last) element in the list of events (whether it's up, down, removed, added)

This would ensure that the latest of up/down (if received) are processed and that any removed or added events are also processed.

adutra replied:
> Shouldn't hostUp be executed right away? It looks like previously only 'schedule' was used for added.

I have been asked to delay UP events as well.

adutra replied:
> There are probably a number of ways to approach this, but the way I think this should work is that on NodeRefreshRequestDeliveryCallback

I would suggest to consider UP/DOWN and ADD/REMOVED as antinomic pairs and only keep the last one of each, if any.

olim7t replied:
> Shouldn't hostUp be executed right away?

According to [CASSANDRA-8236](https://issues.apache.org/jira/browse/CASSANDRA-8236), the problem exists for both UP and ADDED notifications, so not delaying onUp was most likely an oversight (I also discussed it with Sylvain).

> Don't we also have a problem with this sequence of events: onAdd onUp

I wasn't aware that this sequence is possible. For me it was only ADDED when a node gets added. onAdd sets the host up and creates the pools, so there is no need to execute an onUp after.
Anyway, if we process an UP notification for a host that is not known yet, it's turned into an ADDED notification (see `if (hostUp == null) {...` in `Cluster.handle`), so things should work out well. I'll try your test.

tolbertam replied:
Excellent, thanks for the explanation about UP being delayed.   The changes look good so far, will play around a bit and see how it goes :)

tolbertam commented on driver-core/src/main/java/com/datastax/driver/core/Cluster.java:
```diff
@@ -2282,6 +2289,234 @@ public void run() {
+                            break;
+                    }
+                }
+                return Futures.allAsList(futures);
```
It looks like futures is never populated, is that a problem?   It doesn't look like we ever wait on NodeRefreshEvents so it's probably not a problem at the moment.

adutra replied:
It's an oversight. I will get it fixed in a separate commit.

olim7t committed:
Make driver-side metadata optional (JAVA-828).

adutra commented on driver-core/src/main/java/com/datastax/driver/core/Cluster.java:
```diff
@@ -2282,6 +2266,255 @@ public void run() {
+                // gossip is up, but that is before the client-side server is up), so we add a delay
+                // (otherwise the connection will likely fail and have to be retry which is wasteful). This
+                // probably should be fixed C* side, after which we'll be able to remove this.
+                final SettableFuture<?> future = SettableFuture.create();
```
Just noticed that CASSANDRA-8236 is marked fixed for version 2.2.0-beta1. Shouldn't we then only apply a delay if version < 2.2.0?

olim7t commented:
Rebased on 2.0 and force-pushed. I've not squashed the commits yet, merely reordered them.

tolbertam commented on driver-core/src/main/java/com/datastax/driver/core/Cluster.java:
```diff
@@ -2349,6 +2334,255 @@ public void run() {
+                                futures.add(execute(hostDown(downHost)));
+                            break;
+                        case REMOVED:
+                            Host removedHost = metadata.getHost(address);
```
Previously this did:

``` java
removeHost(metadata.getHost(tpAddr), false);
```

Now it will do `hostRemoved` which refreshesNodeInfo first, the node will be gone from peers, so it skips calling onRemove altogether since it is no longer in peers.   I think this is a straightforward fix so I'll commit it, but feel free to back out if you disagree.

olim7t replied:
Yes I agree, we didn't try to refresh the node info before removing before.

tolbertam committed:
Add tests with disabled metadata (JAVA-828).

tolbertam committed:
Disable debouncing for some tests that need metadata immediately.

Also changed ordering of PreparedStatementTest to execute long
repreprepare tests last since they may temporarily break the control
connection causing subsequent tests to fail.

tolbertam commented:
Things are looking pretty good to me right now :+1:.  Had a few remaining comments, but they are minor things.   Did some test cleanup and Jenkins running clean except for 3 tests which I can't get to fail locally (I suspect it's timing oriented, will keep digging into it but shouldn't inhibit merging this) and another test that is caused by a bug in Scassandra with 1.2.19 which i'll fix.

tolbertam commented:
Removed a test that was no longer relevant since setMetadata enabled no longer directly does a nodelist refresh.  Other than that all other tests pass, lgtm if you want to start the merge process, :+1:
